src/lib/vasi-sync/src/lazy_lock.rs

```diff
@@ -0,0 +1,176 @@
+    #[cfg(not(loom))]
+    pub const fn const_new(initializer: Init) -> Self {
+        Self {
+            value: UnsafeCell::new(MaybeUninit::uninit()),
+            init_state: AtomicI8::new(InitState::Uninitd as i8),
+            initializer: UnsafeCell::new(Some(initializer)),
+        }
+    }
+
+    pub fn new(initializer: Init) -> Self {
+        Self {
+            value: UnsafeCell::new(MaybeUninit::uninit()),
+            init_state: AtomicI8::new(InitState::Uninitd.into()),
+            initializer: UnsafeCell::new(Some(initializer)),
+        }
+    }
+
+    fn init(&self, mut state: InitState) {
+        loop {
+            match state {
+                InitState::Uninitd => {
+                    match self.init_state.compare_exchange(
+                        InitState::Uninitd.into(),
+                        InitState::Initializing.into(),
+                        atomic::Ordering::Acquire,
+                        atomic::Ordering::Relaxed,
+                    ) {
+                        Ok(_) => {
+                            // We have the initialization lock.
+                            self.value.get_mut().with(|p: *mut MaybeUninit<T>| unsafe {
+                                let initializer =
+                                    self.initializer.get_mut().deref().take().unwrap();
+                                (*p).write(initializer.initialize());
+                            });
+                            state = InitState::Initd;
+                            self.init_state
+                                .store(state.into(), atomic::Ordering::Release);
+                            break;
+                        }
+                        Err(v) => {
+                            // It changed out from under us; update and try again.
+                            state = v.try_into().unwrap();
+                        }
+                    }
+                }
+                InitState::Initializing => {
+                    // Another thread is initializing. Spin.
+                    // TODO: block?
+                    rustix::process::sched_yield();
+                    state = InitState::try_from(self.init_state.load(atomic::Ordering::Acquire))
+                        .unwrap();
+                }
+                InitState::Initd => {
+                    break;
+                }
+            }
+        }
+    }
+
+    /// Force initialization and return a reference object.
+    ///
+    /// std's LazyLock returns an actual reference here and implements Deref.
+    /// We can't because of our usage of loom's UnsafeCell, which doesn't support
+    /// getting a raw reference that's not borrowed from an intermediate temp object.
+    #[inline]
+    pub fn force(&self) -> Ref<T, Init> {
+        let state = InitState::try_from(self.init_state.load(atomic::Ordering::Acquire)).unwrap();
+        if state != InitState::Initd {
+            self.init(state);
+        }
```

Comment:
My thinking here was to not mark the `init` function `#[inline]` since it's relatively complex, and then try to avoid calling it on the fast path. After further thought I think it probably makes sense to just mark it `#[inline]` too, call it here unconditionally, and let the compiler decide how much inlining to actually do.